### PR TITLE
[MRG] stc plotting tutorial

### PR DESCRIPTION
**Pull request by jasmainak**
This is just to declare my intent ...

I'll finish the PR tomorrow

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
Good to have this.

**Comment by jasmainak** [comment]
Okay, so I think I have a first version ready here: https://9480-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html

**Comment by jasmainak** [comment]
Actually I think this is good to go from my end. Unless I missed something ...

File: tutorials/plot_visualize_stc.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+###############################################################################
+# This time, we have a different container
+# (:class:`VolSourceEstimate <mne.VolSourceEstimate>`) for the source time
+# course.
+print(type(stc))
+
+###############################################################################
+# This too comes with a convenient plot method.
+
+stc.plot(src, subject='sample', subjects_dir=subjects_dir)
```
Comment (agramfort) [inline tutorials/plot_visualize_stc.py]: please also illustrate the matplotlib backend

File: tutorials/plot_visualize_stc.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+###############################################################################
+# The SourceEstimate object is in fact a *surface* source estimate. MNE also
+# supports volume-based source estimates but more on that later.
+#
+# We can plot the source estimate using the
+# :func:`stc.plot <mne.SourceEstimate.plot>` just as in other MNE
+# objects. Note that for this visualization to work, you must have ``mayavi``
+# and ``pysurfer`` installed on your machine.
+initial_time = 0.1
+stc.plot(subjects_dir=subjects_dir, initial_time=initial_time)
```
Comment (agramfort) [inline tutorials/plot_visualize_stc.py]: please also illustrate the matplotlib backend

File: tutorials/plot_visualize_stc.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+##############################################################################
+# Dipoles are fit independantly for each time point, so let us crop our time
+# series to visualize the dipole fit for the time point of interest.
+evoked.crop(0.1, 0.1)
+dip = mne.fit_dipole(evoked, fname_cov, fname_bem, fname_trans)[0]
+
+##############################################################################
+# Finally, we can visualize the dipole.
+dip.plot_locations(fname_trans, 'sample', subjects_dir, mode='orthoview')
```
Comment (agramfort) [inline tutorials/plot_visualize_stc.py]: anything else than orthoview to showcase?

**Comment by jasmainak** [comment]
@agramfort I updated the tutorial to add `matplotlib` viz and also removed the argument `"orthoview"` since that's the only one that's supported. However, the `mayavi` and the `matplotlib` figure look quite different to me.

mayavi
matplotlib

The colorbar is missing + the brain looks flipped. Is this okay for you?

**Comment by agramfort** [comment]
Not ok ... can you look into it ?

**Comment by jasmainak** [comment]
@agramfort I fixed the issue of different views. See https://github.com/mne-tools/mne-python/pull/5528/commits/9cf68682080aaff4a4f85e26a5fa873b4a8218f4 The colormap is still not equivalent. To fix that, I think we should first add a colorbar (which is not currently supported with mpl backend) and then play around with the colormap. Do you want me to do this in this PR itself?

**Comment by jasmainak** [comment]
see https://9509-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html

**Comment by jasmainak** [comment]
Build failures are unrelated to PR

**Comment by agramfort** [comment]
@jasmainak please yes do it in this PR as we can directly see how nice it is. thanks heaps

**Comment by jasmainak** [comment]
Alright. Getting closer. [Colorbar added](https://9522-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html). Not sure why the maps are still not the same though ...

**Comment by larsoner** [comment]
Those maps look the same to me, aside from expected blending and decimation differences. The colorbar rendering in Mayavi is broken but that's not new to this PR.

**Comment by jasmainak** [comment]
@larsoner if you say that the colorbar rendering of mayavi is not correct, then we are fine. This is what I see on my local machine:

comparison

Ignore the fact that the brain maps don't look exactly the same -- it's related to the transparency issue on mayavi that I have. However, the colorbar still seems a bit different, don't you think? The mayavi colorbar seems to squeeze the red part of the "hot" colormap more than the matplotlib colorbar.

I tried a couple of things but no success. Using the pysurfer private function, [_scale_sequential_lut](https://github.com/nipy/PySurfer/blob/master/surfer/viz.py#L2768), I was able to get a similar colorbar but then the brainmap looks different (and full of yellow). So, either you are correct that `mayavi` figure is broken. Or there is still something to fix ...

**Comment by agramfort** [comment]
nice ! can you fix the CIs issues?

**Comment by larsoner** [comment]
Ahh I see what you mean.

You should be able to use the new unified function, it should already do the `cmap` "squeezing". My guess is that in the `mpl` code it is being squeezed again, which is not what we want.

File: mne/viz/_3d.py [hunk 4]
```diff
@@ -1552,6 +1568,20 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
...
+                               norm=plt.Normalize(ctrl_pts[0], ctrl_pts[2]))
+    cax = inset_axes(ax, width="80%", height="5%", loc=8, borderpad=3.)
+    plt.setp(plt.getp(cax, 'xticklabels'), color='w')
+    sm.set_array(np.linspace(0, 1))
+    if colorbar:
+        cb = plt.colorbar(sm, cax=cax, orientation='horizontal')
+        cb_yticks = plt.getp(cax, 'yticklabels')
+        plt.setp(cb_yticks, color='w')
+        cax.tick_params(labelsize=16)
+        cb.patch.set_facecolor((0., 0., 0., 1.0))
```
Comment (larsoner) [inline mne/viz/_3d.py]: Use middle gray (`'0.5'`) not black

File: mne/viz/_3d.py [hunk 5]
```diff
@@ -1460,31 +1460,47 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
...
+                 'fro': {'elev': 16.739, 'azim': 60},
+                 'par': {'elev': 30, 'azim': -60}}
+    kwargs = dict(lh=lh_kwargs, rh=rh_kwargs)
+    if views not in lh_kwargs:
+        raise ValueError("views must be one of ['lat', 'med', 'ros', 'cau', "
                          "'dor' 'ven', 'fro', 'par']. Got %s." % views)
     ctrl_pts, colormap, _, _ = _limits_to_control_points(
-        clim, stc.data, colormap, transparent=False, fmt='matplotlib')
+        clim, stc.data, colormap, transparent=transparent, fmt='matplotlib')
+    colormap = plt.get_cmap(colormap)
```
Comment (larsoner) [inline mne/viz/_3d.py]: You shouldn't need to do this anymore, the output should already be a `cmap` after my recent changes

File: mne/viz/_3d.py [hunk 6]
```diff
@@ -1460,31 +1460,47 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
...
+
+    locs = np.clip(np.linspace(ctrl_pts[0] / ctrl_pts[2], 1, 256), 0, 1)
+    colormap = colormap(locs)
+    colormap = colors.ListedColormap(colormap)
```
Comment (larsoner) [inline mne/viz/_3d.py]: You also should not need to do this, the `_limits_to_control_points` should already do the correct nonlinear scaling

Comment (jasmainak) [reply mne/viz/_3d.py]: I tried the following:

```py

    ctrl_norm = ctrl_pts - ctrl_pts[0]
    ctrl_norm /= ctrl_norm[2]
    vals = np.interp(np.linspace(0, 1, 256), ctrl_norm, [0., 0.5, 1.])
    colormap = np.array(colormap(vals))
    if transparent:
        colormap[:, 3] = np.interp(np.linspace(0, 1, 256), ctrl_norm,
                                   [0, 1, 1])
    colormap = colors.ListedColormap(colormap)
```

but for some strange reason, the brainmap starts saturating then:

brainmap

I can't seem to get both the colorbar and brainmap to look good ...

File: mne/viz/_3d.py (hunk 4 above)
Comment (jasmainak) [reply mne/viz/_3d.py]: It seems Pysurfer uses black. Don't we want to replicate that?

File: mne/viz/_3d.py [hunk 7]
```diff
@@ -1552,6 +1571,21 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
         time_viewer.subplots_adjust(left=0.12, bottom=0.05, right=0.75,
                                     top=0.95)
     fig.subplots_adjust(left=0., bottom=0., right=1., top=1.)
+
+    # add colorbar
+    from mpl_toolkits.axes_grid1.inset_locator import inset_axes
+    sm = plt.cm.ScalarMappable(cmap=cmap,
+                               norm=plt.Normalize(0, ctrl_pts[2]))
```
Comment (jasmainak) [inline mne/viz/_3d.py]: okay I _think_ I got it to work. But there is a hack here that I don't understand at the moment. There is probably a better way to do this ...

File: mne/viz/_3d.py (hunk 4 above)
Comment (larsoner) [reply mne/viz/_3d.py]: It doesn't on master, it uses mid gray (or at least is coded to do so)

**Comment by agramfort** [comment]
@jasmainak please fix the CIs thanks

File: mne/viz/_3d.py [hunk 8]
```diff
@@ -1441,12 +1440,12 @@ def _smooth_plot(this_time, params):
...
-    facecolors[to_blend, :3] = ((1 - transp) *
-                                greymap(curv_ave[to_blend])[:, :3] +
-                                transp * cmap(color_ave[to_blend])[:, :3])
-    facecolors[~to_blend, :3] = greymap(curv_ave[~to_blend])[:, :3]
+    colors = cmap(color_ave)
+    # alpha blend
+    colors[:, :3] *= colors[:, [3]]
+    colors[:, :3] += greymap(curv_ave)[:, :3] * (1.  - colors[:, [3]])
+    colors[:, 3] = 1.
+    facecolors[:] = colors
```
Comment (jasmainak) [inline mne/viz/_3d.py]: @larsoner can you explain what you fixed?

Comment (larsoner) [reply mne/viz/_3d.py]: The alpha blending was wrong before

Comment (larsoner) [reply mne/viz/_3d.py]: (This is now equivalent to standard OpenGL blending like ONE_MINUS_SRC_ALPHA)

**Comment by jasmainak** [comment]
@larsoner are you done? The pysurfer and matplotlib maps still look somewhat different though ...

pysurfer

matplotlib

**Comment by jasmainak** [comment]
oh wait, and now there is no colorbar in matplotlib ...

**Comment by larsoner** [comment]
No I am not done

**Comment by larsoner** [comment]
Should work now. Locally I see this:

screenshot from 2018-10-01 16-21-04

screenshot from 2018-10-01 16-21-06

**Comment by larsoner** [comment]
Updated examples:

https://9547-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html

https://9547-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer_volume.html

https://9547-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_mne_dspm_source_localization.html

**Comment by jasmainak** [comment]
@larsoner the CIs are failing and for the same reason the tutorial is failing on my computer. I don't have `master` of pysurfer ...

**Comment by jasmainak** [comment]
But the visualizations look great now !

**Comment by agramfort** [comment]
looks great but we have now these mayavi colorbar issues.

do you see them too?

**Comment by larsoner** [comment]
Not locally. They are on every PR and in `master` and probably a pain to fix (CI reconfiguration / MESA bug?) so probably shouldn't be handled here.

**Comment by agramfort** [comment]
ok then. Besides the CIs issues LGTM.

thx heaps @larsoner

**Comment by larsoner** [comment]
I'll fix the CIs in a few hours

**Comment by agramfort** [comment]
merge when green.

thx @jasmainak @larsoner

**Pull request merged**